### ISSUE TITLE
Focus Atoms with async atoms?

### ISSUE THREAD

**Issue by mhcoffin** [pmndrs/jotai#231]
Focus atoms seem incredibly cool, but they require a primitive base atom. Unless I'm missing something, that means I lose all the nice asynchronous initialization, right? Is there any way to get both?
 
-mike

**Comment by merisbahti** [comment]
Mey Mike,

Let me see if I understand what you mean, so if we have an atom that's initialized using the `async` constructor like this:

```ts
const asyncAtom  = atom(async () => {
  const response = await Promise.resolve('just an example')
  return response
})
```

Right?

Which results in basically an `Atom<string>`, which is read-only.

So the problem is that `focusAtom` is only able to work on `Primitive` atoms, i.e: `Atom<string, SetState<String>>`.

The reason I did not expand the `focusAtom` to work on read-only-atoms (or async for that matter), is that the value that `focusAtom` provides is that it _both_ sets a getter and a setter. 

But it would be pretty easy to expand the type definition to work on read-only (or async) atoms!

is that kinda what you mean?

**Comment by mhcoffin** [comment]
No, it's not read-only atoms that I want (although for orthogonality that seems like it might be a good idea, too). 

I created a [sandbox](https://codesandbox.io/s/focusatom-with-non-primitive-atoms-dyxqv?file=/src/App.tsx) that shows what I think would be very nice to have. Suspense would take care of the delay in resolving the promise, but the atom could still be updated later. Obviously, for optics to make sense, the atom would have to be WritableAtom<T, T>, with the update type the same as the base type.

What actually happens is that attempting to increment the age sets the original atom's value undefined, I guess because there is a set() function defined. Naively, I would have expected the focus atom to create the updated value using optics and then call 'set(baseAtom, newValue)', but maybe that can't work?

**Comment by merisbahti** [comment]
Oh, I think I understand what you mean, but the sandbox does (for me) not currently contain any async atoms:
![image](https://user-images.githubusercontent.com/2946728/101957256-64e07e80-3c01-11eb-8e16-9aa620e55bcb.png)

So maybe I'll take a look at the sandbox for more detail before I ponder more upon this, but it _sounds_ to me that you have a valid case here, it's just that I didn't consider Read/Write async atoms when I wrote the optics package, but very interesting case and I think we should support it if we can!

**Comment by mhcoffin** [comment]
Hmmm. I thought [sandbox](https://codesandbox.io/s/focusatom-with-non-primitive-atoms-dyxqv?file=/src/App.tsx) saved automagically, but maybe not. Anyway, I just saved again.

**Comment by merisbahti** [comment]
awesome, thanks for the clear example.

this definately should be supported IMO, let me take a look!

**Comment by dai-shi** [comment]
@merisbahti 
#233 will change the `PrimitiveAtom<Value>` type.
We want to accept `WritableAtom<Value, SetStateAction<Value>>` in focusAtom, don't we?

**Comment by merisbahti** [comment]
> @merisbahti
> #233 will change the `PrimitiveAtom<Value>` type.
> We want to accept `WritableAtom<Value, SetStateAction<Value>>` in focusAtom, don't we?

That's right. Or any alias that you decide to call the new `WritableAtom<Value, SetStateAction<Value>>` type.




> No, it's not read-only atoms that I want (although for orthogonality that seems like it might be a good idea, too).
> 
> I created a [sandbox](https://codesandbox.io/s/focusatom-with-non-primitive-atoms-dyxqv?file=/src/App.tsx) that shows what I think would be very nice to have. Suspense would take care of the delay in resolving the promise, but the atom could still be updated later. Obviously, for optics to make sense, the atom would have to be WritableAtom<T, T>, with the update type the same as the base type.
> 
> What actually happens is that attempting to increment the age sets the original atom's value undefined, I guess because there is a set() function defined. Naively, I would have expected the focus atom to create the updated value using optics and then call 'set(baseAtom, newValue)', but maybe that can't work?

Ok so I tested around and was hoping to find a juicy bug or something... Unfortunately no such blessing...

![image](https://user-images.githubusercontent.com/2946728/101978464-73588580-3c55-11eb-8d31-6a8a228d862d.png)

The problem had nothing to do with focusAtom, it's that the personAtom has no "state", so we need to create a "personBaseAtom" which contains the state, then resolve that state using promises in a derived atom which is our `personAtom`

**Comment by merisbahti** [comment]
The exact same behavior can be created by removing `Promise.resolve` from the original sample:
![image](https://user-images.githubusercontent.com/2946728/101978548-0abdd880-3c56-11eb-800e-a48ea5a05e44.png)

(I don't know why codesandbox is saying there's an error in my code...)

**Comment by merisbahti** [comment]
Also, I think it's a good idea to make a test-case that asserts this behavior.

I made one here:
https://github.com/pmndrs/jotai/pull/236

**Comment by dai-shi** [comment]
> Or any alias that you decide to call the new `WritableAtom<Value, SetStateAction<Value>>` type.

I thinks it's explicit and fine for now. It's a bit long though.

**Comment by dai-shi** [comment]
https://github.com/pmndrs/jotai/issues/231#issuecomment-743700759
#239 reverted this change and my previous comment is no longer valid.

This issue #231 is still valid, and we consider it a bug. Either we support async initialization, or if it can't be supported, change the type to restrict it.

**Comment by dai-shi** [comment]
>  `WritableAtom<Value, SetStateAction<Value>>`

After working on splitAtom, I'm sure we can make this to `WritableAtom<Value, Value>`. I will make a small PR to fix it.

The original post is about supporting "async" atoms, which is not solved yet. @Thisen are you going to work on it eventually?

**Comment by Thisen** [comment]
@dai-shi I will, but it's my lowest priority task. Should it bumped up and done after devtools work?

**Comment by dai-shi** [comment]
Yeah, please tackle this after devtools.
If possible, we want to add it before v1. If it's too hard, we give up and can work after v1.

**Comment by Thisen** [comment]
I'm deep into devtools, so if anyone takes to pick this up, please do. If no one picks it up, I'll return to it. :)

**Comment by Thisen** [comment]
@merisbahti @mhcoffin @dai-shi, are there still an issue here? Please fill me in.
A codesandbox reproduction or failing test would be appreciated.

**Comment by dai-shi** [comment]
Okay, I checked the codesandbox in https://github.com/pmndrs/jotai/issues/231#issuecomment-743437912, which is a failing reproduction.

The `personAtom` definition was not correct and fixed:
```js
const personAtom = atom(
  Promise.resolve({ name: "Barney", age: 23 }),
  (get, set, update) => set(personAtom, update)
);
```

Confirmed the bug in v0.11.3, and confirmed it's fixed in v0.16.8.
https://codesandbox.io/s/focusatom-with-non-primitive-atoms-forked-c5d7r?file=/src/App.tsx:440-605

---
Digging it more.
It doesn't work with v0.16.0, but v0.16.1 is fine.
So, it looks like it's fixed by #422 (somewhat by accidentally).
Closing this as resolved.

**Comment by dai-shi** [comment]
btw, `atom(value, write)` usage is not documented. i've been wondering about it. so far it's a hidden feature and only exposes as `atomWithReducer`.

---
another thing found is `const a = atom(() => value, (_, set, v) => { set(a, v) })` is not supported, but seems working by chance. We should probably restrict it.
